File: java/java-practise/src/main/java/day/six/java-cucumber-healthcare/src/test/java/bdd/steps/PatientSearchSteps.java
package bdd.steps;

import domain.Patient;
import domain.PatientDirectory;
import io.cucumber.datatable.DataTable;
import io.cucumber.java.en.And;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;

import java.time.LocalDate;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;

public class PatientSearchSteps {
    private PatientDirectory patientDirectory;
    private List<Patient> searchResults;

    @Given("the patient directory contains:")
    public void thePatientDirectoryContains(
            DataTable table
    ) {
        patientDirectory = new PatientDirectory();

        List<Map<String, String>> rows =
                table.asMaps(String.class, String.class);

        for (Map<String, String> row : rows) {
            Patient patient = new Patient(
                    row.get("id"),
                    row.get("firstName"),
                    row.get("lastName"),
                    LocalDate.parse(
                            row.get("dateOfBirth")
                    )
            );

            patientDirectory.addPatient(patient);
        }
    }

    @When(
            "I search for patients with last name {string}"
    )
    public void iSearchForPatientsWithLastName(
            String lastName
    ) {
        searchResults =
                patientDirectory.searchByLastName(lastName);
    }

    @Then("{int} patients should be returned")
    public void patientsShouldBeReturned(
            int expectedCount
    ) {
        assertThat(searchResults)
                .hasSize(expectedCount);
    }

    @And(
            "the returned patient IDs should be {string}"
    )
    public void theReturnedPatientIdsShouldBe(
            String commaSeparatedIds
    ) {
        List<String> expectedIds =
                Arrays.stream(
                                commaSeparatedIds.split(",")
                        )
                        .map(String::trim)
                        .toList();

        assertThat(searchResults)
                .extracting(Patient::id)
                .containsExactlyElementsOf(expectedIds);
    }
}

File: scharp/day/six/CSharpSpecFlowHealthcareExample/Bindings/PatientSearchSteps.cs

using CSharpSpecFlowHealthcareExample.Domain;
using NUnit.Framework;
using System.Globalization;
using TechTalk.SpecFlow;

namespace CSharpSpecFlowHealthcareExample.Bindings
{
    [Binding]
    public sealed class PatientSearchSteps
    {
        private PatientDirectory _patientDirectory = null!;

        private IReadOnlyList<Patient> _searchResults =
            Array.Empty<Patient>();

        [Given(@"^the patient directory contains:$")]
        public void GivenThePatientDirectoryContains(
            Table table)
        {
            _patientDirectory = new PatientDirectory();

            foreach (var row in table.Rows)
            {
                var patient = new Patient(
                    Id: row["id"],
                    FirstName: row["firstName"],
                    LastName: row["lastName"],
                    DateOfBirth: DateTime.ParseExact(
                        row["dateOfBirth"],
                        "yyyy-MM-dd",
                        CultureInfo.InvariantCulture));

                _patientDirectory.AddPatient(patient);
            }
        }

        [When(
            @"^I search for patients with last name ""([^""]*)""$")]
        public void WhenISearchForPatientsWithLastName(
            string lastName)
        {
            _searchResults =
                _patientDirectory.SearchByLastName(lastName);
        }

        [Then(@"^(\d+) patients should be returned$")]
        public void ThenPatientsShouldBeReturned(
            int expectedCount)
        {
            Assert.That(
                _searchResults,
                Has.Count.EqualTo(expectedCount));
        }

        [Then(
            @"^the returned patient IDs should be ""([^""]*)""$")]
        public void ThenTheReturnedPatientIdsShouldBe(
            string commaSeparatedIds)
        {
            var expectedIds = commaSeparatedIds
                .Split(
                    ',',
                    StringSplitOptions.TrimEntries |
                    StringSplitOptions.RemoveEmptyEntries);

            var actualIds = _searchResults
                .Select(patient => patient.Id)
                .ToArray();

            Assert.That(
                actualIds,
                Is.EqualTo(expectedIds));
        }
    }
}
